share/src/main/java/org/apache/kafka/server/share/persister/PersisterStateManager.java

```diff
@@ -0,0 +1,965 @@
+        }
+
+        long backOff() {
+            return this.backoff.backoff(attempts);
+        }
+    }
+
+    public enum RPCType {
+        READ,
+        WRITE,
+        DELETE,
+        SUMMARY,
+        UNKNOWN
+    }
+
+    public PersisterStateManager(KafkaClient client, ShareCoordinatorMetadataCacheHelper cacheHelper, Time time, Timer timer) {
+        if (client == null) {
+            throw new IllegalArgumentException("Kafkaclient must not be null.");
+        }
+        this.time = time == null ? Time.SYSTEM : time;
+        this.sender = new SendThread(
+            "PersisterStateManager",
+            client,
+            30_000,  //30 seconds
+            this.time,
+            true,
+            new Random(this.time.milliseconds()));
+        this.timer = timer;
+        this.cacheHelper = cacheHelper;
+    }
+
+    public void enqueue(PersisterStateManagerHandler handler) {
+        this.sender.enqueue(handler);
+    }
+
+    public void start() {
+        if (isStarted.compareAndSet(false, true)) {
+            this.sender.start();
+            isStarted.set(true);
+        }
+    }
+
+    public void stop() throws Exception {
+        if (isStarted.compareAndSet(true, false)) {
+            this.sender.shutdown();
+            Utils.closeQuietly(this.timer, "PersisterStateManager timer");
+        }
+    }
+
+    // test visibility
+    Map<Node, Map<RPCType, Map<String, List<PersisterStateManagerHandler>>>> nodeRPCMap() {
+        return nodeRPCMap;
+    }
+
+    public void setGenerateCallback(Runnable generateCallback) {
+        this.generateCallback = generateCallback;
+    }
+
+    /**
+     * Parent class of all RPCs. Uses template pattern to implement core methods.
+     * Various child classes can extend this class to define how to handle RPC specific
+     * responses, retries, batching etc.
+     * <p>
+     * Since the find coordinator RPC/lookup is a necessary pre-condition for all
+     * share state RPCs, the infra code for it is encapsulated in this class itself.
+     */
+    public abstract class PersisterStateManagerHandler implements RequestCompletionHandler {
+        protected Node coordinatorNode;
+        protected final String groupId;
+        protected final Uuid topicId;
+        protected final int partition;
+        private final BackoffManager findCoordBackoff;
+        protected final Logger log = LoggerFactory.getLogger(getClass());
+        private Consumer<ClientResponse> onCompleteCallback;
+
+        public PersisterStateManagerHandler(
+            String groupId,
+            Uuid topicId,
+            int partition,
+            long backoffMs,
+            long backoffMaxMs,
+            int maxRPCRetryAttempts
+        ) {
+            this.groupId = groupId;
+            this.topicId = topicId;
+            this.partition = partition;
+            this.findCoordBackoff = new BackoffManager(maxRPCRetryAttempts, backoffMs, backoffMaxMs);
```

Comment:
A few fields are not initialized here (implicit null). Can we explicitly initialize them here?

Reply:
We can, however I am depending on coordinatorNode to be null as it can be a valid node or noNode, based on the FIND_COORD response. I need a value to signify that we have not yet sent a FIND_CORD request.